tests/cpp/test_multidevice_pipeline.cpp

```diff
@@ -304,6 +304,117 @@ INSTANTIATE_TEST_SUITE_P(
         ::testing::Values(true),
         ::testing::Bool()));
 
+enum class SchedulingMode {
+  InterDeviceOnly,
+  Manual,
+};
+
+std::ostream& operator<<(std::ostream& out, const SchedulingMode& mode) {
+  switch (mode) {
+    case SchedulingMode::InterDeviceOnly:
+      return out << "InterDeviceOnly";
+    case SchedulingMode::Manual:
+      return out << "Manual";
+    default:
+      NVF_ERROR(false);
+  }
+  return out;
+}
+class PipelineTestStagedReduction
+    : public PipelineTest,
+      public ::testing::WithParamInterface<SchedulingMode> {};
+
+// 1D staged reduction
+// Inputs: X[A,B,C]
+TEST_P(PipelineTestStagedReduction, staged_reduction) {
+  auto scheduling_mode = GetParam();
+
+  int num_devices = communicator->size();
+  int A = num_devices;
+  int B = 8;
+  int C = 64;
+  std::vector<int64_t> unsharded_input_sizes = {A, B, C};
+  std::vector<int64_t> input_sizes(unsharded_input_sizes);
+  input_sizes[0] = 1;
+
+  FusionGuard fg(fusion.get());
+  TensorView* tv0 = makeConcreteTensor(unsharded_input_sizes);
+  TensorView* tv1 = sum(tv0, {2});
+  TensorView* tv_out = sum(tv1, {0});
+  fusion->addInput(tv0);
+  fusion->addOutput(tv_out);
+
+  // multi device scheduling:
+  std::vector<int64_t> devices(num_devices);
+  std::iota(devices.begin(), devices.end(), 0);
+  DeviceMesh mesh(devices);
+  for (auto tv : {tv0, tv1, tv_out}) {
+    tv->setDeviceMesh(mesh);
+  }
+  for (auto tv : {tv0, tv1}) {
+    tv->axis(0)->parallelize(ParallelType::DIDx);
+  }
+
+  // Intra-device reduction scheduling for the first reduction:
+  switch (scheduling_mode) {
+    case SchedulingMode::InterDeviceOnly:
+      break;
+    case SchedulingMode::Manual: {
+      // inspired from NVFuserTest.FusionReduction1_CUDA
+      // tv0[I0{A}, I1{B}, I2{C}]
+      tv1->split(2, 32);
+      // tv1[I0{A}, I1{B}, R2o{C/128}, R2i{128}] = tv0[I0{A}, I1{B}, I2{C}]
```

Comment:
IIUC, the comment is not consistent with the code. I believe you intended to say
```suggestion
      // tv1[I0{A}, I1{B}, R2o{C/32}, R2i{32}] = tv0[I0{A}, I1{B}, I2{C}]
```
If so, we should fix other comments too.

Reply:
right